Products and Services

rindo789 opened the pull request:
- created new inline tables in forms of the Deal and the Lead model that distinguish between services and products, referenced in #123 
- created a controller for calculating prices

rindo789 committed:
products and services

wai-blue commented on apps/community/Products/Models/RecordManagers/Product.php:
```diff
@@ -19,4 +19,15 @@ public function SUPPLIER(): HasOne
+    $query = parent::prepareReadQuery($query, $level);
+
+    $main = \ADIOS\Core\Helper::getGlobalApp();
+    if ($main->urlParamAsBool("getServices") == true) $query = $query->where("type", 2);
```
Do not use numbers, use constants.

wai-blue commented on apps/community/Leads/Models/Lead.php:
```diff
@@ -13,6 +13,7 @@
 use HubletoApp\Community\Contacts\Models\Person;
 use HubletoApp\Community\Customers\Models\Customer;
 use HubletoApp\Community\Deals\Models\Deal;
+use HubletoApp\Community\Products\Controllers\API\CalculatePrice;
```
API -> Api

wai-blue commented on apps/community/Products/Controllers/API/CalculatePrice.php:
```diff
@@ -0,0 +1,48 @@
+class CalculatePrice {
+
+  // SINGLE PRODUCT FUNCTIONS
+  public function calculatePriceWithAmount(float $unitPrice, float $amount): float {
```
calculatePriceWithAmount -> calculateFullPrice

wai-blue commented on apps/community/Products/Controllers/API/CalculatePrice.php:
```diff
@@ -0,0 +1,48 @@
+    return $sumPrice * (1 + $tax / 100);
+  }
+
+  public function calculateDiscount(float $sumPrice, float $discount): float {
```
calculateDiscount -> calculateDiscountedPrice

wai-blue replied:
sumPrice -> fullPrice

wai-blue commented on apps/community/Products/Controllers/API/CalculatePrice.php:
```diff
@@ -0,0 +1,48 @@
+
+  public function calculatePriceAfterTax(float $unitPrice, float $amount, float $tax = 0, float $discount = 0): float {
+    $sumPrice = $this->calculatePriceWithAmount($unitPrice, $amount);
+    if ($discount > 0) $sumPrice = $this->calculateDiscount($sumPrice, $discount);
```
remove if $discount

wai-blue commented on apps/community/Products/Controllers/API/CalculatePrice.php:
```diff
@@ -0,0 +1,48 @@
+  }
+
+  public function calculateTax(float $sumPrice, float $tax): float {
+    return $sumPrice * (1 + $tax / 100);
```
return $price * $tax / 100;

wai-blue commented on apps/community/Products/Controllers/API/CalculatePrice.php:
```diff
@@ -0,0 +1,48 @@
+    return $sumPrice;
+  }
+
+  public function calculatePriceAfterTax(float $unitPrice, float $amount, float $tax = 0, float $discount = 0): float {
```
calculatePriceIncludingVat:
return calculatePriceExcludingVat + calculateVat

wai-blue commented on apps/community/Leads/Models/RecordManagers/Lead.php:
```diff
@@ -57,9 +57,22 @@ public function TAGS(): HasMany {
+  public function SERVICES(): HasMany {
+    return $this->hasMany(LeadProduct::class, 'id_lead', 'id')
+        ->whereHas("PRODUCT", function ($query) {
+          $query->where('type', 2);
```
use constants

rindo789 committed:
price calculation adjustments

rindo789 commented:
- changed Tax to Vat in various places for the products and services
- created constants for the product types
- changed some logic in the price calculation, names of functions and the formula of the tax calculation
